scripts/asr_language_modelling/ngram_lm/eval_beamsearch_ngram.py

```diff
@@ -0,0 +1,186 @@
+    )
+    asr_model.preprocessor.featurizer.dither = 0
+    asr_model.preprocessor.featurizer.pad_to = 0
+    # Set model to inference mode
+    asr_model.eval()
+
+    model_tokenizer = asr_model.tokenizer
+    vocabs = list(model_tokenizer.tokenizer.get_vocab().keys())
+
+    target_transcripts = []
+    with open(args.input_manifest, 'r') as manifest_file:
+        audio_file_paths = []
+        for line in tqdm(manifest_file, desc=f"Reading Manifest {args.input_manifest} ..."):
+            data = json.loads(line)
+            target_transcripts.append(data['text'])
+            audio_file_paths.append(data['audio_filepath'])
+
+    # drop it later
+    # audio_file_paths = audio_file_paths[0:10]
+
+    if args.use_probs_cache and os.path.exists(args.probs_cache_file):
+        logging.info(f"Found a pickle file of probabilities at {args.probs_cache_file}.")
+        logging.info(f"Loading the cached pickle file of probabilities from {args.probs_cache_file}...")
+        with open(args.probs_cache_file, 'rb') as probs_file:
+            all_probs = pickle.load(probs_file)
+
+        if len(all_probs) != len(audio_file_paths):
+            raise ValueError(
+                f"The number of samples in the probabilities file '{args.probs_cache_file}' is not "
+                f"the same as the manifest file. "
+                f"You may need to delete the probabilities cached file."
+            )
+    else:
+        with torch.no_grad():
+            all_logits = asr_model.transcribe(audio_file_paths, batch_size=args.acoustic_batch_size, logprobs=True)
+        all_probs = [softmax(logits) for logits in all_logits]
+        logging.info(f"Writing pickle files of probabilities at {args.probs_cache_file}")
+        with open(args.probs_cache_file, 'wb') as f_dump:
+            pickle.dump(all_probs, f_dump)
+
+    wer_dist_greedy = 0
+    words_count = 0
+    for batch_idx, probs in enumerate(all_probs):
+        preds = np.argmax(probs, axis=1)
+        preds_tensor = torch.tensor(preds, device='cpu').unsqueeze(0)
+        pred_text = asr_model._wer.ctc_decoder_predictions_tensor(preds_tensor)[0]
+        pred_split = pred_text.split()
+        target_split = target_transcripts[batch_idx].split()
+        dist = editdistance.eval(target_split, pred_split)
+        wer_dist_greedy += dist
+        words_count += len(target_split)
+
+    # delete the model to free the memory
+    del asr_model
+
+    if args.decoding_mode == "beamsearch_ngram":
+        if not os.path.exists(args.kenlm_model_file):
+            raise FileNotFoundError(f"Could not find the KenLM model file '{args.kenlm_model_file}'.")
+        lm_path = args.kenlm_model_file
+    else:
+        lm_path = None
+
+    if args.decoding_mode in ["beamsearch_ngram", "beamsearch"]:
+        # creating the beam search decoder
+        beam_search_lm = nemo_asr.modules.BeamSearchDecoderWithLM(
+            vocab=[chr(idx + TOKEN_OFFSET) for idx in range(len(vocabs))],
```

Comment:
TODO: Add comments as to this unicode space encoding of subword ids

Reply:
Added